client/verta/verta/dataset/_dataset.py

```diff
@@ -93,14 +93,106 @@ def _set_commit_and_blob_path(self, commit, blob_path):
+        components_to_download = dict()
+
+        for component_blob in self._path_component_blobs:
+            if component_blob.path == component_path:  # exact match with file
+                if implicit_download_to_path:
+                    # default to filename from `component_path` in cwd
+                    local_path = os.path.basename(component_path)
+
+                    # avoid collision with existing file
+                    while os.path.exists(local_path):
+                        local_path = _file_utils.increment_path(local_path)
+                else:
+                    local_path = download_to_path
+
+                return ({component_blob.path: local_path}, local_path)
+            elif component_blob.path.startswith(component_path_as_dir):
+                if implicit_download_to_path:
+                    # default to path relative to parent of `component_path`
+                    #     For example:
+                    #     component_blob.path = "coworker/downloads/data/info.csv"
+                    #     component_path      = "coworker/downloads"
+                    #     local_path          =          "downloads/data/info.csv"
+                    local_path = os.path.relpath(
+                        component_blob.path,
+                        pathlib2.Path(component_path).parent,
+                    )
+
+                    # TODO: move following steps outside of loop if less expensive
+                    path_parts = pathlib2.Path(local_path).parts
+
+                    # user probably doesn't want an s3:/ dir
+                    if path_parts[0] == "s3:":
+                        path_parts = path_parts[1:]
+
+                    # avoid collision with existing directory
+                    root_dirname = path_parts[0]
+                    while os.path.exists(root_dirname):
+                        root_dirname = _file_utils.increment_path(root_dirname)
+                    local_path = os.path.join(root_dirname, *path_parts[1:])
+                else:
+                    # rebase from `component_path` onto `download_to_path`
+                    #     For example:
+                    #     component_blob.path = "coworker/downloads/data/info.csv"
+                    #     component_path      = "coworker/downloads"
+                    #     download_to_path    =            "my-data"
+                    #     local_path          =            "my-data/data/info.csv"
+                    local_path = os.path.join(
+                        download_to_path,
+                        os.path.relpath(component_blob.path, component_path),
+                    )
+
+                components_to_download[component_blob.path] = local_path
+
+        if not components_to_download:
+            raise KeyError("no components found for path {}".format(component_path))
+
+        # determine the root path everything is being downloaded to
```

Comment:
I'm not sure if the segment of code below is properly covered in the tests, but probably because I don't understand the line below. I'll re-check later.

Reply:
It is not covered in tests! Will do